core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -185,7 +185,9 @@ object HostedPartition {
    * This state only applies to brokers that are using a Raft-based metadata
    * quorum; it never happens when using ZooKeeper.
    */
-  final case class Deferred(partition: Partition) extends NonOffline
+  final case class Deferred(partition: Partition,
+                            isNew: Boolean,
+                            onLeadershipChange: (Iterable[Partition], Iterable[Partition]) => Unit) extends NonOffline
```

Comment:
I don't think this callback is logically part of the HostedPartition object itself.  We don't want a different callback for each partition.  It will the same one for each (in practice it just hooks into the group manager and one other manager that I forget).  So let's just supply this callback in whatever function it is needed in, not store it in the hostedpartition object.